java-checks/src/main/java/org/sonar/java/symexec/SymbolicEvaluator.java

```diff
@@ -178,190 +232,129 @@ final boolean isSuperOrThisMemberSelect(ExpressionTree tree) {
-      super.visitArrayAccessExpression(tree);
-      currentResult = SymbolicBooleanConstraint.UNKNOWN;
+      currentStates = nextStates;
     }
 
     @Override
-    public final void visitAssignmentExpression(AssignmentExpressionTree tree) {
-      super.visitAssignmentExpression(tree);
-      Symbol.VariableSymbol symbol = extractVariableSymbol(tree.variable());
+    public final void visitIdentifier(IdentifierTree tree) {
+      Symbol.VariableSymbol symbol = extractVariableSymbol(tree);
       if (symbol != null) {
-        SymbolicValue.SymbolicVariableValue variable = new SymbolicValue.SymbolicVariableValue(symbol);
-        currentState.invalidateRelationsOnValue(variable);
-        currentState.setBooleanConstraint(variable, currentResult);
+        PackedStatementStates nextStates = new PackedStatementStates(currentStates);
+        for (ExecutionState executionState : currentStates) {
+          nextStates.addState(executionState.push(executionState.getValue(symbol)));
+        }
+        currentStates = nextStates;
+      } else {
+        SymbolicValue value = new SymbolicValue();
+        for (ExecutionState executionState : currentStates) {
+          executionState.push(value);
+        }
       }
     }
 
     @Override
-    public void visitIdentifier(IdentifierTree tree) {
-      Symbol.VariableSymbol symbol = extractVariableSymbol(tree);
-      currentResult = symbol != null ? currentState.getBooleanConstraint(new SymbolicValue.SymbolicVariableValue(symbol)) : SymbolicBooleanConstraint.UNKNOWN;
-    }
-
-    @Override
     public final void visitInstanceOf(InstanceOfTree tree) {
-      currentResult = SymbolicBooleanConstraint.UNKNOWN;
+      SymbolicValue result = new SymbolicValue();
+      PackedStatementStates nextStates = new PackedStatementStates(currentStates);
+      for (ExecutionState executionState : currentStates) {
+        for (ExecutionState operandState : evaluateWithState(executionState, tree.expression())) {
+          operandState.pop();
+          nextStates.addState(operandState.push(result));
+        }
+      }
+      currentStates = nextStates;
     }
 
     @Override
     public final void visitLiteral(LiteralTree tree) {
-      if ("false".equals(tree.value())) {
-        currentResult = SymbolicBooleanConstraint.FALSE;
+      SymbolicValue result;
+      if (tree.is(Tree.Kind.INT_LITERAL, Tree.Kind.LONG_LITERAL)) {
+        Long value = LiteralUtils.longLiteralValue(tree);
+        result = value != null ? new SymbolicValue.SymbolicLongValue(value) : new SymbolicValue();
+      } else if ("false".equals(tree.value())) {
+        result = SymbolicValue.BOOLEAN_FALSE;
       } else if ("true".equals(tree.value())) {
-        currentResult = SymbolicBooleanConstraint.TRUE;
+        result = SymbolicValue.BOOLEAN_TRUE;
       } else {
-        currentResult = SymbolicBooleanConstraint.UNKNOWN;
+        result = new SymbolicValue();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this instantiation of "SymbolicValue". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2440)